File: graph_algorithms/maildelivery.cpp
#include <iostream>
#include <vector>
#include <unordered_set>
#include <unordered_map>
#include <stack>
#include <algorithm>

using namespace std;

class DSU {
private:
    vector<int> parent, size;

public:
    DSU(int n) {
        parent.resize(n + 1);
        size.resize(n + 1);
        for (int i = 0; i <= n; i++) {
            parent[i] = i;
            size[i] = 1;
        }
    }

    int find(int a) {
        if (a == parent[a]) {
            return a;
        }
        return parent[a] = find(parent[a]);
    }

    bool union_sets(int a, int b) {
        a = find(a);
        b = find(b);

        if (a == b) {
            return false;
        }

        if (size[a] < size[b]) {
            swap(a, b);
        }

        parent[b] = a;
        size[a] += size[b];
        return true;
    }

    int get_size(int a) {
        return size[find(a)];
    }
};

void dfs(vector<unordered_set<int>>& graph, int node) {
    while(!graph[node].empty()) {
        int neighbor = *(graph[node]).begin();
        graph[node].erase(neighbor);
        graph[neighbor].erase(node);
        dfs(graph, neighbor);
    }
    cout << node << " ";
}

bool is_euler_path_possible(vector<int>& degree) {
    for (int i = 1; i < degree.size(); i++) {
        if (degree[i] % 2 != 0) {
            return false;
        }
    }
    return true;
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n, m;
    cin >> n >> m;

    vector<unordered_set<int>> graph(n + 1);
    DSU dsu(n);

    vector<int> degree(n + 1, 0);
    for (int i = 0; i < m; i++) {
        int a, b;
        cin >> a >> b;
        graph[a].insert(b);
        graph[b].insert(a);
        degree[a]++;
        degree[b]++;

        dsu.union_sets(a, b);
    }

    int components = 0;
    unordered_set<int> root;
    for (int i = 1; i <= n; i++) {
        int val = dsu.find(i);
        if ((!root.count(val) && dsu.get_size(i) > 1) || (i == 1)) {
            components++;
            root.insert(val);
        }
    }

    if (!is_euler_path_possible(degree) || components > 1) {
        cout << "IMPOSSIBLE" << endl;
        return 0;
    }

    dfs(graph, 1);

    return 0;
}

File: graph_algorithms/MailDelivery.java
import java.util.*;
import java.io.*;



public class MailDelivery {

    public static void main(String[] args) {
        FastScanner sc = new FastScanner();
        int n = sc.nextInt();
        int m = sc.nextInt();

        List<Set<Integer>> graph = new ArrayList<>();
        for(int i=0; i<=n; i++) {
            graph.add(new HashSet<>());
        }

        DSU dsu = new DSU(n);

        int degree[] = new int[n+1];
        for(int i=0; i<m; i++) {
            int a = sc.nextInt();
            int b = sc.nextInt();
            graph.get(a).add(b);
            graph.get(b).add(a);
            degree[a]++;
            degree[b]++;

            dsu.union(a, b);
        }

        int components = 0;
        Set<Integer> root = new HashSet<>();
        for(int i=1; i<=n; i++) {
            int val = dsu.find(i);
            if((!root.contains(val) && dsu.getSize(i) > 1) || (i == 1)) {
                components++;
                root.add(val);
            }
        }

        if(!isEulerPathPossible(degree) || components > 1) {
            System.out.println("IMPOSSIBLE");
            return;
        }

        StringBuilder s = new StringBuilder();
        dfs(graph, s, 1);
        System.out.println(s);
    }

    public static void dfs(List<Set<Integer>> graph, StringBuilder s, int node) {
        while(graph.get(node).size() > 0) {
            int val = -1;
            for(int num : graph.get(node)) {
                val = num;
                break;
            }
            graph.get(node).remove(val);
            graph.get(val).remove(node);
            dfs(graph, s, val);
        }
        s.append(node + " ");
    }

    public static boolean isEulerPathPossible(int degree[]) {
        for(int i=1; i<degree.length; i++) {
            if(degree[i]%2 == 1) {
                return false;
            }
        }

        return true;
    }
    
}

class DSU {
    
    private int parent[];
    private int size[];

    public DSU(int n) {
        parent = new int[n+1];
        size = new int[n+1];
        for(int i=0; i<=n; i++) {
            parent[i] = i;
            size[i] = 1;
        }
    }

    public int find(int a) {
        if(a == parent[a]) {
            return a;
        }

        parent[a] = find(parent[a]);
        return parent[a];
    }

    public boolean union(int a, int b) {
        a = find(a);
        b = find(b);

        if(a == b) {
            return false;
        }

        if(size[a] < size[b]) {
            int temp = a;
            a = b;
            b = temp;
        }

        parent[b] = a;
        size[a] += size[b];
        return true;
    }

    public int getSize(int a) {
        return size[find(a)];
    }
}

class FastScanner {
    BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    StringTokenizer st = new StringTokenizer("");
 
    String next() {
        while (!st.hasMoreTokens())
            try {
                st = new StringTokenizer(br.readLine());
            } catch (IOException e) {
                e.printStackTrace();
            }
        return st.nextToken();
    }
 
    int nextInt() {
        return Integer.parseInt(next());
    }
 
    int[] readArray(int n) {
        int[] a = new int[n];
        for (int i = 0; i < n; i++)
            a[i] = nextInt();
        return a;
    }
 
    long nextLong() {
        return Long.parseLong(next());
    }
}

